### DISCUSSION TITLE
TextArea: make sure to use \n as line separator

### DISCUSSION THREAD

**Discussion by UlrichGoebel** [Textualize/textual#4560] [Q&A] [answered]
Hi,

### Short version:

How to make sure, that `TextArea` uses `\n` as line separator, as linux normaly does, regardless of the text loeded and even if no text is loadet?

### Long version:

I read in the [doc](https://textual.textualize.io/widgets/text_area/#line-separators) that the `TextArea` Widget decides from the given content which line separator to use. As I understand, the line separator is one of `\n`, `\r`, `\n\r`, `\r\n`, and this behaviour should make shure, that the text follows the same convention after as before beeing edited.

I am building a UI for existing data in a database. Other UIs may have used different line separators, and I don't have control over that. I know that I have to solve this at database level. Nevertheless I want to make sure, that `TextArea` will use `\n` as line seperator.

Should I scan the text before loading it into `TextArea` and replace `\r`, `\n\r`, `\r\n` by `\n`? Or can I manipulate the `TextArea.document.newline` attribute? Or should I scan/replace after the text has beeing edited?

And which line separator `TextArea` will use if no text is initially loadet?

Any hint is welcome.

Regards
Ulrich

**Comment by TomJGooding** [comment, marked as answer]
It sounds like you already have a pretty good understanding of the `TextArea` widget from the docs.

When content is loaded into the TextArea, the lines of the document are stored separate from the newline character, which is based on the "first occurrence of a line separator". Here's the underlying code (though note only intended for internal use from the leading underscore) :

https://github.com/Textualize/textual/blob/4adaf03050393497949f5282cdbde8d0f7d57d20/src/textual/document/_document.py#L203-L210

You can't change the `TextArea.document.newline` since this is only a property, so I think the best method is to replace any unwanted line separators (at least the first occurance!) before loading into the TextArea.

Since the TextArea doesn't support mixed line endings as all will be converted to this `newline`, you shouldn't need to worry about converting after this content has been edited.

**Reply by UlrichGoebel** [reply]
Thank's, that makes it clear.
